File: VillagesTown/Engines/AIEngine.swift
//
//  AIEngine.swift
//  VillagesTown
//
//  Created by Claude Code
//

import Foundation
import CoreGraphics

class AIEngine {
    let buildingEngine = BuildingConstructionEngine()
    let recruitmentEngine = RecruitmentEngine()
    let movementEngine = MovementEngine()
    let combatEngine = CombatEngine()

    // MARK: - AI Turn
    func executeAITurn(for player: Player, map: inout Map) {
        print("\n🤖 AI Turn: \(player.name) (\(player.aiPersonality?.description ?? "Unknown"))")

        guard !player.isHuman else { return }

        // Get AI villages
        var villages = map.villages.filter { $0.owner == player.id }

        if villages.isEmpty {
            print("   No villages remaining")
            return
        }

        // 1. Economic Phase - Build buildings
        print("   💰 Economic Phase")
        for i in 0..<villages.count {
            makeEconomicDecisions(player: player, village: &villages[i])
        }

        // Update map villages
        for village in villages {
            if let index = map.villages.firstIndex(where: { $0.id == village.id }) {
                map.villages[index] = village
            }
        }

        // 2. Military Phase - Recruit units
        print("   ⚔️ Military Phase")
        villages = map.villages.filter { $0.owner == player.id }
        for i in 0..<villages.count {
            makeMilitaryDecisions(player: player, village: &villages[i], map: &map)
        }

        // Update map villages again
        for village in villages {
            if let index = map.villages.firstIndex(where: { $0.id == village.id }) {
                map.villages[index] = village
            }
        }

        // 3. Combat Phase - Move and attack
        print("   🎯 Combat Phase")
        executeCombatStrategy(player: player, map: &map)
    }

    // MARK: - Economic AI
    func makeEconomicDecisions(player: Player, village: inout Village) {
        guard village.canBuildMore else { return }

        let personality = player.aiPersonality ?? .balanced

        // Determine what to build based on personality
        var buildingPriorities: [Building] = []

        switch personality {
        case .aggressive:
            // Focus on military and resources for military
            buildingPriorities = [
                Building.ironMine,
                Building.barracks,
                Building.market,
                Building.archeryRange
            ]

        case .economic:
            // Focus on economy and infrastructure
            buildingPriorities = [
                Building.farm,
                Building.market,
                Building.lumberMill,
                Building.granary,
                Building.temple,
                Building.ironMine,
                Building.barracks
            ]

        case .balanced:
            // Mix of everything
            buildingPriorities = [
                Building.farm,
                Building.market,
                Building.ironMine,
                Building.barracks,
                Building.lumberMill,
                Building.archeryRange
            ]
        }

        // Try to build first affordable building
        for building in buildingPriorities {
            let check = buildingEngine.canBuild(building: building, in: village)
            if check.can {
                if buildingEngine.buildBuilding(building: building, in: &village) {
                    print("      🏗️ \(village.name): Built \(building.name)")
                    break
                }
            }
        }
    }

    // MARK: - Military AI
    func makeMilitaryDecisions(player: Player, village: inout Village, map: inout Map) {
        let personality = player.aiPersonality ?? .balanced
        let globalResources = GameManager.shared.getGlobalResources(playerID: player.id)

        // Determine recruitment strategy based on global resources
        let shouldRecruit: Bool
        let aggressiveness: Int

        switch personality {
        case .aggressive:
            shouldRecruit = (globalResources[.gold] ?? 0) > 100
            aggressiveness = 3 // Recruit more units

        case .economic:
            shouldRecruit = (globalResources[.gold] ?? 0) > 500
            aggressiveness = 1 // Recruit fewer units

        case .balanced:
            shouldRecruit = (globalResources[.gold] ?? 0) > 300
            aggressiveness = 2 // Moderate recruitment
        }

        guard shouldRecruit else { return }

        // Get available units
        let availableUnits = recruitmentEngine.getAvailableUnits(for: village)

        if availableUnits.isEmpty { return }

        // Choose best unit to recruit
        let unitPriorities: [Unit.UnitType]
        switch personality {
        case .aggressive:
            unitPriorities = [.swordsman, .archer, .militia]
        case .economic:
            unitPriorities = [.spearman, .archer, .militia]
        case .balanced:
            unitPriorities = [.swordsman, .archer, .militia]
        }

        // Try to recruit
        for unitType in unitPriorities {
            if availableUnits.contains(unitType) {
                let quantity = min(aggressiveness, village.population / 10)
                let check = recruitmentEngine.canRecruit(unitType: unitType, quantity: quantity, in: village)

                if check.can {
                    // RecruitmentEngine now handles adding to armies internally
                    let units = recruitmentEngine.recruitUnits(
                        unitType: unitType,
                        quantity: quantity,
                        in: &village,
                        at: village.coordinates
                    )

                    if !units.isEmpty {
                        print("      ⚔️ \(village.name): Recruited \(quantity) \(unitType.rawValue)")
                        break
                    }
                }
            }
        }
    }

    // MARK: - Combat AI (Army-based)
    func executeCombatStrategy(player: Player, map: inout Map) {
        let personality = player.aiPersonality ?? .balanced
        let game = GameManager.shared

        // EARLY GAME PROTECTION: Don't attack before turn 12
        // This gives players time to build up defenses
        let gracePeriod: Int
        switch personality {
        case .aggressive: gracePeriod = 8
        case .economic: gracePeriod = 15
        case .balanced: gracePeriod = 12
        }

        if game.currentTurn < gracePeriod {
            print("      [Grace period - no attacks until turn \(gracePeriod)]")
            return
        }

        // Get AI armies that are stationed (not marching)
        let stationedArmies = game.getStationedArmiesFor(playerID: player.id)

        if stationedArmies.isEmpty {
            print("      No stationed armies")
            return
        }

        // Find enemy targets
        let enemies = map.villages.filter { $0.owner != player.id }

        if enemies.isEmpty {
            print("      No enemy targets")
            return
        }

        // For each stationed army, decide whether to send it
        for army in stationedArmies {
            guard let stationedAtID = army.stationedAt,
                  let stationedAt = map.villages.first(where: { $0.id == stationedAtID }) else {
                continue
            }

            // Find nearest enemy
            var nearestEnemy: Village?
            var shortestDistance = Int.max

            for enemy in enemies {
                let distance = calculateDistance(from: stationedAt.coordinates, to: enemy.coordinates)
                if distance < shortestDistance {
                    shortestDistance = distance
                    nearestEnemy = enemy
                }
            }

            guard let target = nearestEnemy else { continue }

            // Get defender strength estimate (including village garrison)
            let defenderArmies = game.getArmiesAt(villageID: target.id).filter { $0.owner == target.owner }
            let defenderArmyStrength = defenderArmies.reduce(0) { $0 + $1.strength }

            // Villages have inherent garrison strength from population
            let garrisonStrength = target.population / 10  // 10 pop = 1 strength
            let villageDefenseBonus = Int(Double(garrisonStrength) * (1.0 + target.defenseBonus))
            let totalDefenderStrength = defenderArmyStrength + villageDefenseBonus

            let attackerStrength = army.strength

            // Decide if should attack based on personality
            // Now requires SIGNIFICANT advantage because of garrison
            let shouldAttack: Bool
            switch personality {
            case .aggressive:
                // Need 1.5x strength to attack (was 0.7x)
                shouldAttack = attackerStrength > 0 && Double(attackerStrength) > Double(totalDefenderStrength) * 1.5
            case .economic:
                // Need 2.5x strength to attack (was 1.5x)
                shouldAttack = attackerStrength > 0 && Double(attackerStrength) > Double(totalDefenderStrength) * 2.5
            case .balanced:
                // Need 2.0x strength to attack (was 1.1x)
                shouldAttack = attackerStrength > 0 && Double(attackerStrength) > Double(totalDefenderStrength) * 2.0
            }

            // Require larger army sizes before attacking
            let minArmySize: Int
            switch personality {
            case .aggressive: minArmySize = 5
            case .economic: minArmySize = 10
            case .balanced: minArmySize = 7
            }

            if shouldAttack && army.unitCount >= minArmySize {
                // Send army to attack!
                if game.sendArmy(armyID: army.id, to: target.id) {
                    print("      🚶 \(army.name) (\(army.unitCount) units) sent to attack \(target.name)")
                    print("         Attacker: \(attackerStrength) vs Defender: \(totalDefenderStrength) (includes garrison)")
                }
            }
        }
    }

    private func calculateDistance(from: CGPoint, to: CGPoint) -> Int {
        let dx = abs(Int(to.x) - Int(from.x))
        let dy = abs(Int(to.y) - Int(from.y))
        return max(dx, dy)
    }
}

File: villages_town_flutter/lib/engines/ai_engine.dart
import 'dart:math';
import 'dart:ui';
import '../data/map/game_map.dart';
import '../data/models/ai_personality.dart';
import '../data/models/building.dart';
import '../data/models/player.dart';
import '../data/models/resource.dart';
import '../data/models/unit_type.dart';
import '../data/models/village.dart';
import 'building_construction_engine.dart';
import 'game_manager.dart';
import 'recruitment_engine.dart';

class AIEngine {
  final BuildingConstructionEngine _buildingEngine = BuildingConstructionEngine();
  final RecruitmentEngine _recruitmentEngine = RecruitmentEngine();

  void executeAITurn(Player player, GameMap map) {
    if (player.isHuman) return;

    var villages = map.villages.where((v) => v.owner == player.id).toList();
    if (villages.isEmpty) return;

    // 1. Economic Phase
    for (final village in villages) {
      _makeEconomicDecisions(player, village);
    }

    // 2. Military Phase
    villages = map.villages.where((v) => v.owner == player.id).toList();
    for (final village in villages) {
      _makeMilitaryDecisions(player, village, map);
    }

    // 3. Combat Phase
    _executeCombatStrategy(player, map);
  }

  void _makeEconomicDecisions(Player player, Village village) {
    if (!village.canBuildMore) return;

    final personality = player.aiPersonality ?? AIPersonality.balanced;

    List<Building> buildingPriorities;
    switch (personality) {
      case AIPersonality.aggressive:
        buildingPriorities = [
          Building.barracks,
          Building.ironMine,
          Building.market,
          Building.archeryRange,
          Building.farm,
        ];
      case AIPersonality.economic:
        buildingPriorities = [
          Building.farm,
          Building.market,
          Building.lumberMill,
          Building.barracks,
          Building.granary,
          Building.temple,
          Building.ironMine,
        ];
      case AIPersonality.balanced:
        buildingPriorities = [
          Building.barracks,
          Building.farm,
          Building.market,
          Building.ironMine,
          Building.lumberMill,
          Building.archeryRange,
        ];
    }

    for (final building in buildingPriorities) {
      if (village.buildings.any((b) => b.name == building.name)) continue;

      final (can, _) = _buildingEngine.canBuild(building, village);
      if (can) {
        _buildingEngine.buildBuilding(building, village);
        GameManager.shared.updateVillage(village);
        break;
      }
    }
  }

  void _makeMilitaryDecisions(Player player, Village village, GameMap map) {
    final personality = player.aiPersonality ?? AIPersonality.balanced;
    final globalResources = GameManager.shared.getGlobalResources(player.id);

    final availableUnits = _recruitmentEngine.getAvailableUnits(village);
    if (availableUnits.isEmpty) return;

    final int goldThreshold;
    final int recruitCount;

    switch (personality) {
      case AIPersonality.aggressive:
        goldThreshold = 50;
        recruitCount = 3;
      case AIPersonality.economic:
        goldThreshold = 200;
        recruitCount = 1;
      case AIPersonality.balanced:
        goldThreshold = 100;
        recruitCount = 2;
    }

    if ((globalResources[Resource.gold] ?? 0) <= goldThreshold) return;

    final unitPriorities = [UnitType.militia, UnitType.swordsman, UnitType.archer, UnitType.spearman];

    for (final unitType in unitPriorities) {
      if (availableUnits.contains(unitType)) {
        final (can, _) = _recruitmentEngine.canRecruit(unitType, recruitCount, village);
        if (can) {
          _recruitmentEngine.recruitUnits(unitType, recruitCount, village, village.coordinates);
          break;
        }
      }
    }
  }

  void _executeCombatStrategy(Player player, GameMap map) {
    final personality = player.aiPersonality ?? AIPersonality.balanced;
    final game = GameManager.shared;

    final int gracePeriod;
    switch (personality) {
      case AIPersonality.aggressive:
        gracePeriod = 5;
      case AIPersonality.economic:
        gracePeriod = 10;
      case AIPersonality.balanced:
        gracePeriod = 7;
    }

    if (game.currentTurn < gracePeriod) return;

    final stationedArmies = game.getStationedArmiesFor(player.id);
    if (stationedArmies.isEmpty) return;

    final enemies = map.villages.where((v) => v.owner != player.id).toList();
    if (enemies.isEmpty) return;

    for (final army in stationedArmies) {
      final stationedAtId = army.stationedAt;
      if (stationedAtId == null) continue;

      final stationedAt = map.villages.cast<Village?>().firstWhere(
            (v) => v!.id == stationedAtId,
            orElse: () => null,
          );
      if (stationedAt == null) continue;

      // Find best target
      Village? bestTarget;
      var bestTargetScore = -1 << 30;

      for (final enemy in enemies) {
        final defenderArmies = game.getArmiesAt(enemy.id).where((a) => a.owner == enemy.owner);
        final defenderArmyStrength = defenderArmies.fold(0, (sum, a) => sum + a.strength);
        final garrisonStrength = enemy.garrisonStrength * 3;
        final totalDefenderStrength = defenderArmyStrength + garrisonStrength;

        final distance = _calculateDistance(stationedAt.coordinates, enemy.coordinates);
        final advantage = army.strength - totalDefenderStrength;
        var score = advantage - (distance * 5);

        // Bonus for neutral villages
        if (enemy.owner == 'neutral') score += 50;

        if (score > bestTargetScore) {
          bestTargetScore = score;
          bestTarget = enemy;
        }
      }

      if (bestTarget == null) continue;

      final defenderArmies = game.getArmiesAt(bestTarget.id).where((a) => a.owner == bestTarget!.owner);
      final defenderArmyStrength = defenderArmies.fold(0, (sum, a) => sum + a.strength);
      final totalDefenderStrength = defenderArmyStrength + bestTarget.garrisonStrength * 3;

      final bool shouldAttack;
      switch (personality) {
        case AIPersonality.aggressive:
          shouldAttack = army.strength > 0 && army.strength > totalDefenderStrength * 0.8;
        case AIPersonality.economic:
          shouldAttack = army.strength > 0 && army.strength > totalDefenderStrength * 1.5;
        case AIPersonality.balanced:
          shouldAttack = army.strength > 0 && army.strength >= totalDefenderStrength;
      }

      final int minArmySize;
      switch (personality) {
        case AIPersonality.aggressive:
          minArmySize = 3;
        case AIPersonality.economic:
          minArmySize = 5;
        case AIPersonality.balanced:
          minArmySize = 4;
      }

      final isUndefendedNeutral =
          bestTarget.owner == 'neutral' && defenderArmyStrength == 0 && bestTarget.garrisonStrength < 5;

      if ((shouldAttack && army.unitCount >= minArmySize) || (isUndefendedNeutral && army.unitCount >= 2)) {
        game.sendArmy(army.id, bestTarget.id);
      }
    }
  }

  int _calculateDistance(Offset from, Offset to) {
    final dx = (to.dx - from.dx).abs().toInt();
    final dy = (to.dy - from.dy).abs().toInt();
    return max(dx, dy);
  }
}
